Source/DafnyCore/AST/Grammar/ParseErrors.cs

```diff
@@ -407,18 +510,562 @@ declared identifier.
+", range => new List<DafnyAction> {
+    OneAction("replace `( )` with `map[ ]`", range, "map[" + range.PrintOriginal()[1..^1] + "]", false),
+    OneAction("replace `( )` with `imap[ ]`", range, "imap[" + range.PrintOriginal()[1..^1] + "]", false),
+});
+
+    Add(ErrorId.p_must_be_multiset,
+    @"
+A set/iset/multiset display expression can have two forms. 
+One form is a list of values enclosed by curly braces: `var c := multiset{1,2,2,3}`.
+The other appears as a conversion operation: `var c := multiset(s)`.
+However, this second form can only be used to convert a set to a multiset.
+
+In the current parser, however, this error message is unreachable, so if it appears
+please report the error.
+The tests that check for this error case are already known to be false by previous testing.
+");
+
+    Add(ErrorId.p_seq_has_one_type_argument,
+    @"
+The built-in `seq` (sequence) type takes one type parameter, which in some situations is inferred.
+That type parameter is the type of the sequence elements.
+"); // TODO - get rid of the extra parameters
+
+    Add(ErrorId.p_map_comprehension_must_have_term_expression,
+    @"
+A map comprehension defines a map, which associates a value with each member of a set of keys.
+The full syntax for a map comprehension looks like `map x | 0 <= x < 5:: x*2 => x*3`
+which maps the keys `0, 2, 4, 6, 8` to the values `0, 3, 6, 9, 12` respectively.
+
+One can abbreviate the above syntax to expressions like `map x | 0 <= x < 5 :: x*3`,
+which is equivalent to `map x | 0 <= x < 5 :: x => x*3`. The missing expression before
+the `=>` is just the declared identifier.
+
+One can also have multiple variables involved as in `map x, y | 0 < x < y < 5 :: 10*x+y => 10*y+x`,
+which defines the mappings `(12=>21, 13=>31, 14=>41, 23=>32, 24=>42, 34=>43)`.
+
+But when there are multiple variables, one cannot abbreviate the `=>` syntax with just its right-hand expression,
+because it is not clear what the left-hand expression should be. 
+
+Incorrect text like `const s := map x, y  | 0 <= x < y < 10 :: x*y` should be written
+ as `const s := map x, y  | 0 <= x < y < 10 :: f(x,y) => x*y` for some `f(x,y)` that gives
+a unique value for each pair of `x,y` permitted by the range expression (here `0 <= x < y < 10`).
+");
+
+    Add(ErrorId.p_no_patterns_in_let_such_that,
+    @"
+The let-such-that expression initializes a variable to some value satisfying a given condition.
+For example, one might write `const cc := var x: int :| x <= 10; x`,
+where `cc` would get some value `x` satisfying `x < 10`.
+
+For simplicity, however, Dafny requires the variables being initialized to be simple names, not patterns.
+");
+
+    Add(ErrorId.p_no_equal_in_let_initialization,
+    @"
+Like local variables, let variables are initialized with `:=` (and sometimes with `:-` or `:|`), but not with `=`.
+In Dafny `=` is used only in type definitions.
+", range => new List<DafnyAction> {
```

Comment:
Can you please reuse the same lambda as for the local variables?

Reply:
Did it - though my preference is to keep the list of actions local.